File: github_cpp/5/72.cpp
#include <iostream>
#include "SortTestHelper.h"
#include "SelectionSort.h"

using namespace std;

template<typename T>
void insertionSort(T arr[], int n){

    for( int i = 1 ; i < n ; i ++ ) {
        for( int j = i ; j > 0 && arr[j] < arr[j-1] ; j -- )
            swap( arr[j] , arr[j-1] );
    }

    return;
}

int main() {

    int n = 10000;
    cout<<"Test for Random Array, size = "<<n<<", random range [0, "<<n<<"]"<<endl;
    int *arr1 = SortTestHelper::generateRandomArray(n,0,n);
    int *arr2 = SortTestHelper::copyIntArray(arr1, n);

    SortTestHelper::testSort("Insertion Sort", insertionSort,arr1,n);
    SortTestHelper::testSort("Selection Sort", selectionSort,arr2,n);

    delete[] arr1;
    delete[] arr2;

    cout<<endl;

    return 0;
}

File: github_java_sort/9/72.java
package sort;


public class Selectionsort {
private int[] arr;
	
	public Selectionsort(int[] a){
		this.arr = a;
		selectionsort(arr, arr.length);
		printArray(arr);
	}
	
	private void selectionsort(int[] a, int len) {
		
		for (int i = 0; i < len-1; i++) {
			int min = i;	
			
			
			for (int j = i+1; j < len; j++) {
				
				if (a[j] <= a[min]) {
					min = j;
				}
			}
			
			
			int temp = a[i];
			a[i] = a[min];
			a[min] = temp;
		}
	}

	private void printArray(int[] a) {
		for (int i = 0; i < a.length; i++) {
			System.out.print(a[i] + " ");
		}
	}
	public static void main(String[] args) {
		int[]x = new int[]{7,2,1,6,9,8,5,3,4};

		Selectionsort ss = new Selectionsort(x);
	}
}
